_posts/2019-01-31-What-you-need-from-your-CI.md

```diff
@@ -0,0 +1,160 @@
+---
+title: What you need from your CI
+excerpt: The main things to focus on regarding a CI. This is our experience with CIEs in the last 7 years.
+layout: post
+date: 2019-01-31 00:00:00 +0100
+author: abtris
+published: true
+---
+
+Apiary is using multiple Continuous Integration Engines (CIE). Our testing tool [Dredd](https://dredd.org/en/latest/), for example, uses [Travis CI](https://travis-ci.org/) to support multiple versions of Node.JS, and [AppVeyor](https://www.appveyor.com/) to support Windows.
+For a couple of projects we use [Wercker](https://www.oracle.com/corporate/acquisitions/wercker/) which supports the Oracle Managed Kubernetes running our workload.
+
+
+While our stack is based on Node.JS, we use many other languages including C++, Python, Ruby and Go. We use Git as our source control management. The Apiary team consists of about 25 engineers working across more than 100 projects. In this post I will focus on the biggest Apiary project, a big monolith app in Node.JS with external services using FaaS (AWS Lambda).
+
+We started using [CircleCI](https://circleci.com/) for our application in 2012. In 2015 we introduced parallelism for tests, initially with 2 nodes which we have since increased to 8. In 2016 we moved our testing to Docker to improve the isolation. Since 2017 we have been using [CircleCI 2.0](https://circleci.com/blog/say-hello-to-circleci-2-0/) (docker native).
+
+[![](/images/2019-01-11-Why-is-Apiary-using-CircleCI/apiary_timeline_small.jpg)](/images/2019-01-11-Why-is-Apiary-using-CircleCI/apiary_timeline.jpg)
+
+
+## What do you need from your CIE?
+
+In my point of view, these are the most important features of a CIE:
+
+1. Speed
+2. Reliability
+3. Scalability
+4. Maintenance costs
+
+
+## 1. Speed
+
+Firstly let me ask this: do you ever think about your builds? Try answering these questions:
+
+- Do you know your time to deploy?
+- Do you know how much time is spent in queue?
+- What is the maximum acceptable time for your build (5, 15, 30 min, 4h)?
+- Do you monitor those times or are you just guessing?
+
+In 2016 we had problems with the speed of our test suite. As you can see on the graph below, our tests in that year took 28 min in May and ended up taking up to 50 min in November. This caused multiple problems and affected the team morale. People started ignoring flaky tests and running more retries. The time to deploy was horrible and we couldn't deploy new features as fast as we built them.
+
+![](/images/2019-01-11-Why-is-Apiary-using-CircleCI/buildtimes.png)
+
+We launched an engineering project to solve the problem of the long time it took to deploy. We started with these main points:
+
+- Awareness
+- Groups
+- Parallelism
+
+We looked for easy ways to improve the results. We invested more money and divided the test suite to run in parallel containers, going from 4 to 8 containers and eventually ended up using 33 containers in CircleCI.
```

Comment:
this can seem a bit confusing - going from 4 to 8 and ending with 33 containers. I'd suggest:

```suggestion
We looked for easy ways to improve the results. We invested more money and divided the test suite to run in parallel containers, going from 4 to 8 containers per single build and eventually ended up using 33 containers in CircleCI (enabling 4 parallel builds of the monolith).
```

Reply:
Good point. Fixed.